src/hotspot/share/opto/block.cpp

```diff
@@ -954,6 +945,52 @@ void PhaseCFG::fixup_flow() {
   } // End of for all blocks
 }
 
+void PhaseCFG::remove_unreachable_blocks() {
+  ResourceMark rm;
+  Block_List unreachable;
+  // Initialize worklist of unreachable blocks to be removed.
+  for (uint i = 0; i < number_of_blocks(); i++) {
+    Block* block = get_block(i);
+    if (block->is_trivially_unreachable()) {
+      unreachable.push(block);
+    }
+  }
+  // Now remove all blocks that are transitively unreachable.
+  while (unreachable.size() > 0) {
+    Block* dead = unreachable.pop();
+    for (uint i = 0; i < _number_of_blocks; i++) {
```

Comment:
Wouldn't we want to iterate backward here and adjust block->_pre_order until we hit dead?

Reply:
Thanks for the suggestion, Roland! Yes, since `get_block(i)->_pre_order == i` holds at this stage, it is enough with iterating over the elements that succeed `dead` in the block list. I have updated the changeset accordingly.